src/main.rs

```diff
@@ -18,25 +18,119 @@ fn main() {
 }
 
 fn real_main() -> Result<i32, Box<dyn std::error::Error>> {
+    let rng = Rng::new();
     let cargo = std::env::var("CARGO").unwrap_or_else(|_| "cargo".to_owned());
     let mut arg_iter = std::env::args().peekable();
-    let _cargo = arg_iter.next();
-    if arg_iter.peek().map_or(false, |arg| arg == "mommy") {
-        let _mommy = arg_iter.next();
+
+    // Understand who mommy really is~
+    //
+    // *INHALES*
+    //
+    // Ok so first argument is this binary, should look like /some/path/to/cargo-mommy(.exe)
+    // but we want to let it instead be cargo-daddy, and for everything to rekey itself to:
+    // * make the default role be "daddy"
+    // * make all the read env-vars be "CARGO_DADDYS_*"
+    let mut platform_uses_exe = false;
+    let raw_first_arg = arg_iter.next().unwrap_or_default();
+    let first_arg = if let Some(stripped) = raw_first_arg.strip_suffix(".exe") {
+        platform_uses_exe = true;
+        stripped.to_owned()
+    } else {
+        raw_first_arg.clone()
+    };
+    let true_role = if let Some((_path, role)) = first_arg.rsplit_once("cargo-") {
+        role.to_owned()
+    } else {
+        // If something messed up is going on, default to "mommy"
+        "mommy".to_owned()
+    };
+
+    // *GASPS FOR BREATH*
+    //
+    // *INHALES DESPERATELY*
+    //
+    // cargo subcommands when run as "cargo blah" get the "blah" argument passed to themselves
+    // as the *second* argument. However if we are invoked directly as "cargo-blah" we won't
+    // have that extra argument! So if there's a second argument that is "mommy" (or whatever)
+    // we pop off that redundant copy before forwarding the rest of the args back to "cargo ...".
+    // if we don't do this, we'll infinitely recurse into ourselves by re-calling "cargo mommy"!
+    // (note that it *is* supported to do `cargo mommy mommy` and get two messages, although I
+    // belive we do this, `cargo-mommy mommy` will still only get you onw message).
+
+    if arg_iter.peek().map_or(false, |arg| arg == &true_role) {
+        let _role = arg_iter.next();
     }
 
+    // *WHEEZES*
+    //
+    // *PANTS FOR A MINUTE*
+    //
+    // Ok so now we want to detect if the invocation looked like "cargo mommy i mean daddy"
+    // if it *does* we want to copy ourselves to rename cargo-mommy to cargo-daddy. We can't
+    // clone std::env::args() and we need to peek more than one element, so we create a
+    // "speculation" buffer that we will chain back into the args on failed speculation.
+    // on successful speculation we'll purge the buffer, effectively consuming the args.
+    //
+    // ...
+    //
+    // ~
+    let mut speculated = vec![];
```

Comment:
Can we just collect the arguments into a vector in the first place? It seems like we're doing a lot of work here to avoid collecting, but it seems to me like collecting into a vector will just result in a lot simpler code without leaving much performance on the table.